### ISSUE TITLE
Tracking of durations for Initialize/Cleanup methods

### ISSUE THREAD

**Issue by Danielku15** [microsoft/testfx#493]
## Description
This is a feature request and not a bug. 

In general our unit tests are rather fast (individually) but still on our CI server the overall testrun takes long. The total run time is ~30mins whereas the sum of all test durations is only half of it. In our case this seems to be mainly caused by the whole `AssemblyInitialize`, `AssemblyCleanup`, `TestInitialize` and `TestCleanup` methods we have in place. 

Unfortunately these durations are not tracked and contained in the test result reports (*.trx) and therefore not shown anywhere. 

It would be great if MSTest would also keep track of the overall, and individual durations for this setup and teardown phases and then also include them in the test report. 

Personally I would prefer this information: 

* Per test the Initialize, Run and Cleanup durations. 
Location: In the `UnitTestResult` tag where we have the duration today
* Per assembly Initialize/Run/Cleanup durations. 
Location: Maybe as extension of the `Times` tag containing sub-tags per assembly.

**Comment by jayaranigarg** [comment]
@Danielku15 : Thank you for the feedback. 
@pvlakshm to look into the ask.

**Comment by AbhitejJohn** [comment]
@Danielku15 : The overall test run time reported in the trx file includes the times takes for setup/teardown as well. However it might not be clearly evident how long these operations take unless someone does a diff with individual test runtimes which isn't very useful. Based on what you are saying it looks like we need some way of limiting the time these setup/teardown operations take. Reporting them would be a +. I was wondering if adding timeouts for Assembly/Class Init/Cleanup would help your scenario.

**Comment by Danielku15** [comment]
> The overall test run time reported in the trx file includes the times takes for setup/teardown as well. 
Does that also apply for Assembly setup/teardown? Are they accounted to the first test running (which is not deterministic)? 

> Based on what you are saying it looks like we need some way of limiting the time these setup/teardown operations take. 
I'm personally more interested into having reporting of the times than timeouts. We have no direct need to limit the time this operations are taking. Even though the time we need there might be long compared to the actual tests, it is a necessary step in our test environment. Especially for our integration tests we have quite some bunch of test data we need to import into our system, and that takes its time. Currently we are rather lacking of the information that the setup/teardown operations are the ones making the overall test execution longer than only the tests themselves.

**Comment by Danielku15** [comment]
I checked a bit what would need to be done for this tracking. Maybe I'll start working on a PR for it so I analyzed a bit how things are currently working. This stuff might be clear for the core devs of this project but for me it helps  😉 

- The AssemblyInitialize is run here: https://github.com/Microsoft/testfx/blob/3197e7bb9f7fd4219c7855be575fb5def205aff7/src/Adapter/MSTest.CoreAdapter/Execution/TestAssemblyInfo.cs#L116
- The AssemblyCleanup is run here: https://github.com/Microsoft/testfx/blob/3197e7bb9f7fd4219c7855be575fb5def205aff7/src/Adapter/MSTest.CoreAdapter/Execution/TestAssemblyInfo.cs#L200
- The ClassInitialize is run here: https://github.com/Microsoft/testfx/blob/3197e7bb9f7fd4219c7855be575fb5def205aff7/src/Adapter/MSTest.CoreAdapter/Execution/TestClassInfo.cs#L220
- The ClassCleanup is run here: https://github.com/Microsoft/testfx/blob/3197e7bb9f7fd4219c7855be575fb5def205aff7/src/Adapter/MSTest.CoreAdapter/Execution/TestClassInfo.cs#L302
- The TestInitialize is run here: https://github.com/Microsoft/testfx/blob/3197e7bb9f7fd4219c7855be575fb5def205aff7/src/Adapter/MSTest.CoreAdapter/Execution/TestMethodInfo.cs#L522
- The TestCleanup is run here: https://github.com/Microsoft/testfx/blob/3197e7bb9f7fd4219c7855be575fb5def205aff7/src/Adapter/MSTest.CoreAdapter/Execution/TestMethodInfo.cs#L419
- The individual test duration is stored here: https://github.com/Microsoft/testfx/blob/bec913906cce0a3e8182ec3b9f82168ab96dbb6c/src/Adapter/MSTest.CoreAdapter/ObjectModel/UnitTestResult.cs#L96
- The individual test duration is calculated here: https://github.com/Microsoft/testfx/blob/3197e7bb9f7fd4219c7855be575fb5def205aff7/src/Adapter/MSTest.CoreAdapter/Execution/TestMethodInfo.cs#L126

Tasks that need to be done:
- [ ] Extend the object model to hold the assembly, class, test init/cleanup durations.
- [ ] Measure the execution times in the exeuction places and store them into the object model 
- [ ] Ensure durations are written into the result files. 
- [ ] Create unit and integration Tests. 

To be defined:
- [ ] Decide if durations should be stored hierarchically and summed up (e.g. on assembly we also have a TestInitializeDuration that contains the sum of all initialize durations within the assembly) This way you would not need to traverse the whole XML to sum up manually.

**Comment by Danielku15** [comment]
I started again to look into this a bit deeper and attempted to implement it. Unfortunately the object model of vstest does not foresee yet any place where we could hold the durations of all these additional timings. 

For AssemblyInitialize and ClassInitialize and TestInitialize I attached additional properties to `UnitTestResult`. If the initialize methods are executed as part of a particular test, the times will be filled there. To get back the hierarchical structure of times, the individual UnitTestResult objects need to be scanned and matched with the TestMethod.
